presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,265 @@
+
+import java.util.Optional;
+
+import static com.facebook.presto.sql.analyzer.MaterializedViewInformationExtractor.MaterializedViewInfo;
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewQueryOptimizer
+        extends AstVisitor<Node, Void>
+{
+    private final Table materializedView;
+    private final Query materializedViewQuery;
+
+    private MaterializedViewInfo materializedViewInfo;
+
+    private static final Logger logger = Logger.get(MaterializedViewQueryOptimizer.class);
+
+    public MaterializedViewQueryOptimizer(Table materializedView, Query materializedViewQuery)
+    {
+        this.materializedView = requireNonNull(materializedView, "materialized view is null");
+        this.materializedViewQuery = requireNonNull(materializedViewQuery, "materialized view query is null");
+    }
+
+    public Node rewrite(Node node)
+    {
+        try {
+            MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
+            materializedViewInformationExtractor.process(materializedViewQuery);
+            this.materializedViewInfo = materializedViewInformationExtractor.getMaterializedViewInfo();
+            return process(node);
+        }
+        catch (Exception ex) {
+            logger.error(ex, "Failed to rewrite query: %s.", node);
+            return node;
+        }
+    }
+
+    @Override
+    protected Node visitNode(Node node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected Node visitQuery(Query node, Void context)
+    {
+        QueryBody rewrittenQueryBody = (QueryBody) process(node.getQueryBody(), context);
+        return new Query(
+                node.getWith(),
+                rewrittenQueryBody,
+                node.getOrderBy(),
+                node.getOffset(),
+                node.getLimit());
+    }
+
+    @Override
+    protected Node visitQuerySpecification(QuerySpecification node, Void context)
+    {
+        if (!node.getFrom().isPresent()) {
+            throw new IllegalStateException("Query with no From clause is not rewritable by materialized view");
+        }
+        Optional<Relation> rewrittenFrom = node.getFrom().map(from -> (Relation) process(from, context));
+        Select rewrittenSelect = (Select) process(node.getSelect(), context);
+
+        // TODO: Handle filter containment problem https://github.com/prestodb/presto/issues/16405
+        if (materializedViewInfo.getWhereClause().isPresent() && !materializedViewInfo.getWhereClause().equals(node.getWhere())) {
+            throw new IllegalStateException("Query with no groupBy clause is not rewritable by materialized view with groupBy clause");
+        }
+        Optional<Expression> rewrittenWhere = node.getWhere().map(where -> (Expression) process(where, context));
+
+        if (materializedViewInfo.getGroupBy().isPresent() && !node.getGroupBy().isPresent()) {
```

Comment:
Same as above, we should be able to move this in the visitGroupBy?

Reply:
@jainxrohit The optimizer will only process visitGroupBy if there exists a GroupBy in the base query. If not, then it will skip the visitGroupBy.

Here I am only trying to eliminate case that does not have groupBy when the MV has groupBy.

Reply:
got it, thanks for explaining!